### ISSUE TITLE
alloca?

### ISSUE THREAD

**Issue by nickdesaulniers** [h2o/h2o#115]
There's many uses of alloca, which is making me feel itchy.  Can we remove them, or at least check that the value that would be passed to them is not above a certain amount?

**Comment by kazuho** [comment]
> There's many uses of alloca, which is making me feel itchy. Can we remove them,

No.  As using `alloca` has two benefits: speed and simplicity (no need to free the allocated memory).

IMO there is nothing wrong with allocating variable length structure on the stack.  In fact, C99 has added support for such feature which is variable-length array.  Unfortunately it is not supported in MSVC, thait is why we need to use `alloca`.

> or at least check that the value that would be passed to them is not above a certain amount?

Generally speaking, when writing a server, you should always have an upper-bound for the size of the memory being allocated.  H2O is designed as such (expect for `h2o_buffer_reserve` which falls back to temporary-file-backed `mmap(2)`).  Memory allocation failures will lead to `abort(3)` (in case of `h2o_mem_alloc`) or SEGV (in case of `alloca(3)`.

So the only case in which you should use `h2o_mem_alloc` in place of `alloca(3)` is when the requested size is fairly large so that it might overflow the stack.  Please open an issue or send a PR when you find any such case.

PS. you'd might be interested in reading http://blog.kazuhooku.com/2014/12/memory-management-in-h2o.html
